File: bindings/kotlin/src/main/kotlin/strling/emitters/Pcre2Emitter.kt
package strling.emitters

import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import strling.core.*

/**
 * STRling PCRE2 Emitter - IR to PCRE2 Pattern String
 *
 * This class implements the emitter that transforms STRling's Intermediate
 * Representation (IR) into PCRE2-compatible regex pattern strings.
 */
object Pcre2Emitter {
    
    /**
     * Escapes PCRE2 metacharacters in literal strings.
     */
    fun escapeLiteral(s: String): String {
        val metaChars = setOf('.', '^', '$', '|', '(', ')', '?', '*', '+', '{', '}', '[', ']', '\\')
        val result = StringBuilder()
        
        for (ch in s) {
            if (ch in metaChars) {
                result.append('\\').append(ch)
            } else {
                result.append(ch)
            }
        }
        
        return result.toString()
    }
    
    /**
     * Escapes a character for use inside [...] per PCRE2 rules.
     */
    fun escapeClassChar(ch: String): String {
        if (ch.length != 1) {
            throw IllegalArgumentException("escapeClassChar expects single character")
        }
        
        val c = ch[0]
        return when (c) {
            '\\', ']' -> "\\$c"
            '[' -> "\\["
            '-' -> "\\-"
            '^' -> "\\^"
            '\n' -> "\\n"
            '\r' -> "\\r"
            '\t' -> "\\t"
            '\u000C' -> "\\f"
            '\u000B' -> "\\v"
            else -> {
                val code = c.code
                when {
                    code < 32 || (code in 127..159) -> String.format("\\x%02x", code)
                    else -> ch
                }
            }
        }
    }
    
    /**
     * Emit a PCRE2 character class.
     */
    private fun emitClass(cc: IRCharClass): String {
        val items = cc.items
        
        // Single-item shorthand optimization
        if (items.size == 1 && items[0] is IRClassEscape) {
            val esc = items[0] as IRClassEscape
            val k = esc.type
            val prop = esc.property
            
            when (k) {
                "d", "w", "s" -> {
                    if (cc.negated) {
                        return when (k) {
                            "d" -> "\\D"
                            "w" -> "\\W"
                            "s" -> "\\S"
                            else -> "\\$k"
                        }
                    }
                    return "\\$k"
                }
                "D", "W", "S" -> {
                    val base = k.lowercase()
                    return if (cc.negated) "\\$base" else "\\$k"
                }
                "p", "P" -> {
                    if (prop != null) {
                        val isUpperP = k == "P"
                        val useUpperP = cc.negated xor isUpperP
                        val use = if (useUpperP) "P" else "p"
                        return "\\$use{$prop}"
                    }
                }
            }
        }
        
        // General case: build a bracket class
        val parts = StringBuilder()
        for (item in items) {
            when (item) {
                is IRClassLiteral -> {
                    parts.append(escapeClassChar(item.char))
                }
                is IRClassRange -> {
                    parts.append(escapeClassChar(item.from))
                    parts.append('-')
                    parts.append(escapeClassChar(item.to))
                }
                is IRClassEscape -> {
                    when {
                        item.type.matches(Regex("[dDwWsS]")) -> {
                            parts.append('\\').append(item.type)
                        }
                        (item.type == "p" || item.type == "P") && item.property != null -> {
                            parts.append('\\').append(item.type).append('{').append(item.property).append('}')
                        }
                        else -> {
                            parts.append('\\').append(item.type)
                        }
                    }
                }
            }
        }
        
        val inner = parts.toString()
        return "[" + (if (cc.negated) "^" else "") + inner + "]"
    }
    
    /**
     * Emit quantifier suffix.
     */
    private fun emitQuantSuffix(minv: Int, maxv: Any, mode: String): String {
        val q = when {
            minv == 0 && maxv == "Inf" -> "*"
            minv == 1 && maxv == "Inf" -> "+"
            minv == 0 && maxv == 1 -> "?"
            minv == maxv -> "{$minv}"
            maxv == "Inf" -> "{$minv,}"
            else -> "{$minv,$maxv}"
        }
        
        return when (mode) {
            "Lazy" -> "$q?"
            "Possessive" -> "$q+"
            else -> q
        }
    }
    
    /**
     * Return true if 'child' needs a non-capturing group when quantifying.
     */
    private fun needsGroupForQuant(child: IROp): Boolean {
        return when (child) {
            is IRCharClass, is IRDot, is IRGroup, is IRBackref, is IRAnchor -> false
            is IRLit -> {
                // Single escape sequences don't need grouping
                if (child.value.length == 2 && child.value[0] == '\\') false
                else child.value.length > 1
            }
            is IRAlt, is IRLook -> true
            is IRSeq -> child.parts.size > 1
            else -> false
        }
    }
    
    /**
     * Generate opening for group based on type.
     */
    private fun emitGroupOpen(g: IRGroup): String {
        return when {
            g.atomic == true -> "(?>"
            g.capturing -> {
                if (g.name != null) "(?<${g.name}>"
                else "("
            }
            else -> "(?:"
        }
    }
    
    /**
     * Default upper bound on AST/IR nesting depth before the emitter
     * aborts. Mirrors the SSOT in the TypeScript reference and the
     * matching constants in every other binding.
     */
    const val DEFAULT_MAX_DEPTH = 250

    private const val REDOS_MESSAGE =
        "The pattern contains overlapping alternations or nested unbounded " +
            "quantifiers (e.g., (a+)+). This can lead to catastrophic backtracking " +
            "and exponential CPU spikes. Consider using possessive quantifiers " +
            "(++ or *+) or atomic groups to guarantee execution safety."

    /**
     * Mutable per-emit context threaded through [emitNode] so the depth,
     * lookbehind and warning-collection guards can fire without
     * polluting the public API.
     */
    private class EmitContext(maxDepth: Int) {
        var depth: Int = 0
        val maxDepth: Int = if (maxDepth > 0) maxDepth else DEFAULT_MAX_DEPTH
        var inLookbehind: Boolean = false
        val warnings: MutableList<STRlingWarning> = mutableListOf()
    }

    /** Resolve a `IRQuant.max` JsonElement to either an Int or the string "Inf". */
    private fun resolveMax(maxEl: Any): Any {
        return when {
            maxEl is JsonPrimitive -> maxEl.intOrNull ?: maxEl.content
            maxEl.toString() == "\"Inf\"" -> "Inf"
            else -> maxEl.toString().replace("\"", "")
        }
    }

    /** True iff a quantifier has an unbounded upper bound. */
    private fun isUnboundedQuant(q: IRQuant): Boolean = resolveMax(q.max) == "Inf"

    /** True iff a quantifier matches a variable number of characters. */
    private fun isVariableLengthQuant(q: IRQuant): Boolean {
        val mx = resolveMax(q.max)
        return mx != q.min
    }

    /**
     * Mirror of `_isFixedLengthBody` in the TS SSOT. Returns true when
     * [node] consumes a fixed number of characters and is therefore
     * safe inside a PCRE2 lookbehind.
     */
    private fun isFixedLengthBody(node: IROp): Boolean = when (node) {
        is IRQuant -> !isVariableLengthQuant(node) && isFixedLengthBody(node.child)
        is IRSeq -> node.parts.all { isFixedLengthBody(it) }
        is IRAlt -> node.branches.all { isFixedLengthBody(it) }
        is IRGroup -> isFixedLengthBody(node.body)
        is IRLook -> true // lookarounds are zero-width
        else -> true
    }

    /**
     * Mirror of `_hasNestedUnboundedQuant` in the TS SSOT. Detects an
     * unbounded quantifier reachable via single-child wrappers or any
     * branch of an Alt.
     */
    private fun hasNestedUnboundedQuant(child: IROp): Boolean = when (child) {
        is IRQuant -> isUnboundedQuant(child)
        is IRGroup -> hasNestedUnboundedQuant(child.body)
        is IRSeq -> child.parts.size == 1 && hasNestedUnboundedQuant(child.parts[0])
        is IRAlt -> child.branches.any { hasNestedUnboundedQuant(it) }
        else -> false
    }

    private fun pushReDoSWarning(ctx: EmitContext) {
        if (ctx.warnings.any { it.code == "REDOS_RISK" }) return
        ctx.warnings.add(STRlingWarning("REDOS_RISK", REDOS_MESSAGE))
    }

    /**
     * Emit a single IR node to PCRE2 syntax.
     */
    private fun emitNode(node: IROp, parentKind: String): String {
        return emitNode(node, parentKind, EmitContext(0))
    }

    private fun emitNode(node: IROp, parentKind: String, ctx: EmitContext): String {
        ctx.depth += 1
        try {
            if (ctx.depth > ctx.maxDepth) {
                throw STRlingCompilationError(
                    "Maximum AST depth exceeded (limit: ${ctx.maxDepth}). " +
                        "This pattern is too deeply nested and risks host stack " +
                        "exhaustion during emission. Refactor the pattern to " +
                        "reduce nesting, or flatten capturing groups where possible.",
                    "MAX_DEPTH",
                )
            }
            return when (node) {
                is IRLit -> escapeLiteral(node.value)
                is IRDot -> "."
                is IRAnchor -> {
                    val at = if (node.at == "NonWordBoundary") "NotWordBoundary" else node.at
                    when (at) {
                        "Start" -> "^"
                        "End" -> "$"
                        "WordBoundary" -> "\\b"
                        "NotWordBoundary" -> "\\B"
                        "AbsoluteStart" -> "\\A"
                        "EndBeforeFinalNewline" -> "\\Z"
                        "AbsoluteEnd" -> "\\z"
                        else -> ""
                    }
                }
                is IRBackref -> {
                    when {
                        node.byName != null -> "\\k<${node.byName}>"
                        node.byIndex != null -> "\\${node.byIndex}"
                        else -> ""
                    }
                }
                is IRCharClass -> emitClass(node)
                is IRSeq -> node.parts.joinToString("") { emitNode(it, "Seq", ctx) }
                is IRAlt -> {
                    val body = node.branches.joinToString("|") { emitNode(it, "Alt", ctx) }
                    if (parentKind in listOf("Seq", "Quant")) "(?:$body)" else body
                }
                is IRQuant -> {
                    // ReDoS guard: only flag when the *outer* quantifier is
                    // itself unbounded (e.g. `(a+)+`). A bounded outer like
                    // `(a+){0,3}` cannot produce exponential backtracking on
                    // its own.
                    if (isUnboundedQuant(node) && hasNestedUnboundedQuant(node.child)) {
                        pushReDoSWarning(ctx)
                    }
                    var childStr = emitNode(node.child, "Quant", ctx)
                    if (needsGroupForQuant(node.child) && node.child !is IRGroup) {
                        childStr = "(?:$childStr)"
                    }
                    childStr + emitQuantSuffix(node.min, resolveMax(node.max), node.mode)
                }
                is IRGroup -> emitGroupOpen(node) + emitNode(node.body, "Group", ctx) + ")"
                is IRLook -> {
                    // Variable-length lookbehind guard: PCRE2 mandates a
                    // fixed-width lookbehind body. Detect the violation here
                    // so the user sees a Signpost-pattern error rather than
                    // an opaque PCRE2 compile failure leaking from the runtime.
                    if (node.dir == "Behind" && !isFixedLengthBody(node.body)) {
                        throw STRlingCompilationError(
                            "PCRE2 does not support variable-length lookbehinds. " +
                                "The lookbehind body contains a quantifier that makes " +
                                "its length unpredictable. Rewrite the assertion using " +
                                "a fixed-length range (e.g. `{1,8}` instead of `+`), " +
                                "or restructure the pattern using a Lookahead, or " +
                                "extract the quantified portion outside the assertion.",
                            "VLB_NOT_SUPPORTED",
                        )
                    }
                    val wasInLb = ctx.inLookbehind
                    if (node.dir == "Behind") ctx.inLookbehind = true
                    try {
                        val op = when {
                            node.dir == "Ahead" && !node.neg -> "?="
                            node.dir == "Ahead" && node.neg -> "?!"
                            node.dir == "Behind" && !node.neg -> "?<="
                            node.dir == "Behind" && node.neg -> "?<!"
                            else -> "?="
                        }
                        "($op${emitNode(node.body, "Look", ctx)})"
                    } finally {
                        ctx.inLookbehind = wasInLb
                    }
                }
            }
        } finally {
            ctx.depth -= 1
        }
    }
    
    /**
     * Build the inline prefix form for flags.
     */
    private fun emitPrefixFromFlags(flags: Flags): String {
        val letters = StringBuilder()
        if (flags.ignoreCase) letters.append('i')
        if (flags.multiline) letters.append('m')
        if (flags.dotAll) letters.append('s')
        if (flags.unicode) letters.append('u')
        if (flags.extended) letters.append('x')
        
        return if (letters.isNotEmpty()) "(?$letters)" else ""
    }
    
    /**
     * Emit a PCRE2 pattern string from IR.
     */
    fun emit(irRoot: IROp, flags: Flags? = null): String {
        return emitWithDiagnostics(irRoot, flags, 0).pattern
    }

    /**
     * Emit a PCRE2 pattern string from IR and surface any non-fatal
     * diagnostics collected during emission. Pass [maxDepth] <= 0 to
     * use [DEFAULT_MAX_DEPTH].
     */
    fun emitWithDiagnostics(irRoot: IROp, flags: Flags? = null, maxDepth: Int = 0): CompileResult {
        val prefix = if (flags != null) emitPrefixFromFlags(flags) else ""
        val ctx = EmitContext(maxDepth)
        val body = emitNode(irRoot, "", ctx)
        return CompileResult(prefix + body, ctx.warnings.toList())
    }
}

File: bindings/php/src/Emitters/Pcre2Emitter.php
<?php

declare(strict_types=1);

namespace STRling\Emitters;

use STRling\Core\Nodes\{
    Node, Alternation, Sequence, Literal, Dot, Anchor, CharacterClass, Quantifier, Group, Backreference,
    Lookahead, NegativeLookahead, Lookbehind, NegativeLookbehind,
    ClassItem, Escape, Range, Flags
};
use STRling\Core\CompileResult;
use STRling\Core\STRlingCompilationError;
use STRling\Core\STRlingWarning;

/**
 * STRling PCRE2 Emitter - IR to PCRE2 Pattern String
 *
 * Transforms STRling AST nodes into PCRE2-compatible regex pattern strings.
 */
class Pcre2Emitter
{
    /**
     * Default upper bound on AST/IR nesting depth before the emitter
     * aborts. Mirrors the SSOT in the TypeScript reference and the
     * matching constants in every other binding.
     */
    public const DEFAULT_MAX_DEPTH = 250;

    private const REDOS_MESSAGE =
        'The pattern contains overlapping alternations or nested unbounded '
      . 'quantifiers (e.g., (a+)+). This can lead to catastrophic backtracking '
      . 'and exponential CPU spikes. Consider using possessive quantifiers '
      . '(++ or *+) or atomic groups to guarantee execution safety.';

    /**
     * Emit a PCRE2 pattern string from an AST node.
     */
    public static function emit(Node $node, ?Flags $flags = null): string
    {
        return self::emitWithDiagnostics($node, $flags, 0)->pattern;
    }

    /**
     * Emit a PCRE2 pattern string from an AST node and surface any
     * non-fatal diagnostics collected during emission. Pass
     * `$maxDepth <= 0` to use {@see self::DEFAULT_MAX_DEPTH}.
     */
    public static function emitWithDiagnostics(Node $node, ?Flags $flags = null, int $maxDepth = 0): CompileResult
    {
        $prefix = $flags ? self::emitFlagsPrefix($flags) : '';
        $ctx = self::newContext($maxDepth);
        $body = self::emitNode($node, '', $ctx);
        return new CompileResult($prefix . $body, $ctx['warnings']);
    }

    /**
     * @return array{depth:int, maxDepth:int, inLookbehind:bool, warnings:list<STRlingWarning>}
     */
    private static function newContext(int $maxDepth): array
    {
        return [
            'depth' => 0,
            'maxDepth' => $maxDepth > 0 ? $maxDepth : self::DEFAULT_MAX_DEPTH,
            'inLookbehind' => false,
            'warnings' => [],
        ];
    }

    /** True iff a quantifier has an unbounded upper bound. */
    private static function isUnboundedQuant(Quantifier $q): bool
    {
        // null in the user-facing AST means unbounded; 'inf' is also accepted.
        return $q->max === null || (is_string($q->max) && strtolower($q->max) === 'inf');
    }

    /** True iff a quantifier matches a variable number of characters. */
    private static function isVariableLengthQuant(Quantifier $q): bool
    {
        if (self::isUnboundedQuant($q)) {
            return true;
        }
        return $q->max !== $q->min;
    }

    /**
     * Mirror of `_isFixedLengthBody` in the TS SSOT. Returns true when
     * the node consumes a fixed number of characters and is therefore
     * safe inside a PCRE2 lookbehind.
     */
    private static function isFixedLengthBody(Node $node): bool
    {
        if ($node instanceof Quantifier) {
            return !self::isVariableLengthQuant($node) && self::isFixedLengthBody($node->target);
        }
        if ($node instanceof Sequence) {
            foreach ($node->parts as $p) {
                if (!self::isFixedLengthBody($p)) return false;
            }
            return true;
        }
        if ($node instanceof Alternation) {
            foreach ($node->alternatives as $b) {
                if (!self::isFixedLengthBody($b)) return false;
            }
            return true;
        }
        if ($node instanceof Group) {
            return self::isFixedLengthBody($node->body);
        }
        if ($node instanceof Lookahead || $node instanceof NegativeLookahead
            || $node instanceof Lookbehind || $node instanceof NegativeLookbehind) {
            return true; // zero-width
        }
        return true;
    }

    /**
     * Mirror of `_hasNestedUnboundedQuant` in the TS SSOT. Detects an
     * unbounded quantifier reachable via single-child wrappers or any
     * branch of an Alternation.
     */
    private static function hasNestedUnboundedQuant(Node $child): bool
    {
        if ($child instanceof Quantifier) {
            return self::isUnboundedQuant($child);
        }
        if ($child instanceof Group) {
            return self::hasNestedUnboundedQuant($child->body);
        }
        if ($child instanceof Sequence) {
            return count($child->parts) === 1 && self::hasNestedUnboundedQuant($child->parts[0]);
        }
        if ($child instanceof Alternation) {
            foreach ($child->alternatives as $b) {
                if (self::hasNestedUnboundedQuant($b)) return true;
            }
            return false;
        }
        return false;
    }

    /** Append a single REDOS_RISK warning, deduplicated per pass. */
    private static function pushReDoSWarning(array &$ctx): void
    {
        foreach ($ctx['warnings'] as $w) {
            if ($w->code === 'REDOS_RISK') return;
        }
        $ctx['warnings'][] = new STRlingWarning('REDOS_RISK', self::REDOS_MESSAGE);
    }

    /**
     * Build the inline prefix form from flags, e.g. "(?imx)"
     */
    private static function emitFlagsPrefix(Flags $flags): string
    {
        $letters = '';
        if ($flags->ignoreCase) $letters .= 'i';
        if ($flags->multiline) $letters .= 'm';
        if ($flags->dotAll) $letters .= 's';
        if ($flags->unicode) $letters .= 'u';
        if ($flags->extended) $letters .= 'x';
        return $letters ? "(?{$letters})" : '';
    }

    private static function emitNode(Node $node, string $parentKind, array &$ctx): string
    {
        $ctx['depth'] += 1;
        try {
            if ($ctx['depth'] > $ctx['maxDepth']) {
                throw new STRlingCompilationError(
                    "Maximum AST depth exceeded (limit: {$ctx['maxDepth']}). "
                  . 'This pattern is too deeply nested and risks host stack '
                  . 'exhaustion during emission. Refactor the pattern to '
                  . 'reduce nesting, or flatten capturing groups where possible.',
                    'MAX_DEPTH'
                );
            }

            if ($node instanceof Literal) {
                return self::escapeLiteral($node->value);
            }

            if ($node instanceof Dot) {
                return '.';
            }

            if ($node instanceof Anchor) {
                return match ($node->at) {
                    'Start' => '^',
                    'End' => '$',
                    'WordBoundary' => '\\b',
                    'NotWordBoundary' => '\\B',
                    'AbsoluteStart' => '\\A',
                    'EndBeforeFinalNewline' => '\\Z',
                    'AbsoluteEnd' => '\\z',
                    default => '',
                };
            }

            if ($node instanceof Backreference) {
                if ($node->name !== null) {
                    return "\\k<{$node->name}>";
                }
                if ($node->index !== null) {
                    return "\\{$node->index}";
                }
                return '';
            }

            if ($node instanceof CharacterClass) {
                return self::emitClass($node);
            }

            if ($node instanceof Sequence) {
                $parts = array_map(fn($p) => self::emitNode($p, 'Seq', $ctx), $node->parts);
                return implode('', $parts);
            }

            if ($node instanceof Alternation) {
                $branches = array_map(fn($b) => self::emitNode($b, 'Alt', $ctx), $node->alternatives);
                $body = implode('|', $branches);
                return in_array($parentKind, ['Seq', 'Quant'], true) ? "(?:{$body})" : $body;
            }

            if ($node instanceof Quantifier) {
                // ReDoS guard: only flag when the *outer* quantifier is
                // itself unbounded (e.g. `(a+)+`). A bounded outer like
                // `(a+){0,3}` cannot produce exponential backtracking on
                // its own.
                if (self::isUnboundedQuant($node) && self::hasNestedUnboundedQuant($node->target)) {
                    self::pushReDoSWarning($ctx);
                }
                $childStr = self::emitNode($node->target, 'Quant', $ctx);
                if (self::needsGroupForQuant($node->target)) {
                    $childStr = "(?:{$childStr})";
                }
                return $childStr . self::emitQuantSuffix($node);
            }

            if ($node instanceof Group) {
                $open = self::emitGroupOpen($node);
                return $open . self::emitNode($node->body, 'Group', $ctx) . ')';
            }

            if ($node instanceof Lookahead) {
                return '(?=' . self::emitNode($node->body, 'Look', $ctx) . ')';
            }

            if ($node instanceof NegativeLookahead) {
                return '(?!' . self::emitNode($node->body, 'Look', $ctx) . ')';
            }

            if ($node instanceof Lookbehind || $node instanceof NegativeLookbehind) {
                // Variable-length lookbehind guard: PCRE2 mandates a
                // fixed-width lookbehind body. Detect the violation here
                // so the user sees a Signpost-pattern error rather than
                // an opaque PCRE2 compile failure leaking from the runtime.
                if (!self::isFixedLengthBody($node->body)) {
                    throw new STRlingCompilationError(
                        'PCRE2 does not support variable-length lookbehinds. '
                      . 'The lookbehind body contains a quantifier that makes '
                      . 'its length unpredictable. Rewrite the assertion using '
                      . 'a fixed-length range (e.g. `{1,8}` instead of `+`), '
                      . 'or restructure the pattern using a Lookahead, or '
                      . 'extract the quantified portion outside the assertion.',
                        'VLB_NOT_SUPPORTED'
                    );
                }
                $wasInLb = $ctx['inLookbehind'];
                $ctx['inLookbehind'] = true;
                try {
                    $op = $node instanceof NegativeLookbehind ? '(?<!' : '(?<=';
                    return $op . self::emitNode($node->body, 'Look', $ctx) . ')';
                } finally {
                    $ctx['inLookbehind'] = $wasInLb;
                }
            }

            throw new \RuntimeException('Emitter missing for ' . get_class($node));
        } finally {
            $ctx['depth'] -= 1;
        }
    }

    private static function escapeLiteral(string $s): string
    {
        $toEscape = [' ', '#', '$', '&', '(', ')', '*', '+', '-', '.', '?', '[', '\\', ']', '^', '{', '|', '}', '~'];
        $result = '';

        for ($i = 0; $i < strlen($s); $i++) {
            $ch = $s[$i];
            if (in_array($ch, $toEscape, true) && $ch !== '-') {
                $result .= '\\' . $ch;
            } else {
                $result .= $ch;
            }
        }

        return $result;
    }

    private static function escapeClassChar(string $ch): string
    {
        if ($ch === '\\' || $ch === ']') {
            return '\\' . $ch;
        }
        if ($ch === '-') {
            return '\\-';
        }
        if ($ch === '^') {
            return '\\^';
        }
        if ($ch === "\n") return '\\n';
        if ($ch === "\r") return '\\r';
        if ($ch === "\t") return '\\t';
        if ($ch === "\f") return '\\f';
        if ($ch === "\v") return '\\v';

        $code = ord($ch);
        if ($code < 32 || ($code >= 127 && $code <= 159)) {
            return sprintf('\\x%02x', $code);
        }

        return $ch;
    }

    private static function emitClass(CharacterClass $cc): string
    {
        $items = $cc->members;

        // Single-item shorthand optimization
        if (count($items) === 1 && $items[0] instanceof Escape) {
            $esc = $items[0];
            $k = match ($esc->kind) {
                'digit' => 'd',
                'not-digit' => 'D',
                'word' => 'w',
                'not-word' => 'W',
                'space' => 's',
                'not-space' => 'S',
                default => null,
            };

            if ($k !== null) {
                if (in_array($k, ['d', 'w', 's'], true)) {
                    if ($cc->negated) {
                        return '\\' . strtoupper($k);
                    }
                    return '\\' . $k;
                }
                if (in_array($k, ['D', 'W', 'S'], true)) {
                    $base = strtolower($k);
                    return $cc->negated ? '\\' . $base : '\\' . $k;
                }
            }
        }

        // General case: build bracket class
        $parts = [];
        foreach ($items as $item) {
            if ($item instanceof Literal) {
                $parts[] = self::escapeClassChar($item->value);
            } elseif ($item instanceof Range) {
                $parts[] = self::escapeClassChar($item->from) . '-' . self::escapeClassChar($item->to);
            } elseif ($item instanceof Escape) {
                $k = match ($item->kind) {
                    'digit' => 'd',
                    'not-digit' => 'D',
                    'word' => 'w',
                    'not-word' => 'W',
                    'space' => 's',
                    'not-space' => 'S',
                    'property' => 'p',
                    'not-property' => 'P',
                    default => null,
                };
                if ($k !== null) {
                    $parts[] = '\\' . $k;
                }
            }
        }

        $inner = implode('', $parts);
        return '[' . ($cc->negated ? '^' : '') . $inner . ']';
    }

    private static function emitQuantSuffix(Quantifier $q): string
    {
        $min = $q->min;
        $max = $q->max;

        if ($min === 0 && $max === null) {
            $suffix = '*';
        } elseif ($min === 1 && $max === null) {
            $suffix = '+';
        } elseif ($min === 0 && $max === 1) {
            $suffix = '?';
        } elseif ($min === $max) {
            $suffix = "{{$min}}";
        } elseif ($max === null) {
            $suffix = "{{$min},}";
        } else {
            $suffix = "{{$min},{$max}}";
        }

        if ($q->lazy) {
            $suffix .= '?';
        } elseif ($q->possessive) {
            $suffix .= '+';
        }

        return $suffix;
    }

    private static function needsGroupForQuant(Node $child): bool
    {
        if ($child instanceof CharacterClass || $child instanceof Dot ||
            $child instanceof Group || $child instanceof Backreference ||
            $child instanceof Anchor) {
            return false;
        }
        if ($child instanceof Literal) {
            return strlen($child->value) > 1;
        }
        if ($child instanceof Alternation) {
            return true;
        }
        if ($child instanceof Sequence) {
            return count($child->parts) > 1;
        }
        return false;
    }

    private static function emitGroupOpen(Group $group): string
    {
        if ($group->atomic === true) {
            return '(?>';
        }
        if ($group->capturing) {
            if ($group->name !== null) {
                return "(?<{$group->name}>";
            }
            return '(';
        }
        return '(?:';
    }
}
